chain/chain/src/chain.rs

```diff
@@ -4771,27 +4771,23 @@ impl Chain {
         receipts: &[Receipt],
         shard_layout: &ShardLayout,
     ) -> Vec<CryptoHash> {
-        if shard_layout.num_shards() == 1 {
+        let num_shards = shard_layout.shard_ids().count();
+        if num_shards == 1 {
             return vec![CryptoHash::hash_borsh(ReceiptList(0, receipts))];
         }
-        let mut account_id_to_shard_id_map = HashMap::new();
-        let mut shard_receipts: Vec<_> =
-            shard_layout.shard_ids().into_iter().map(|shard_id| (shard_id, Vec::new())).collect();
-        for receipt in receipts.iter() {
-            let shard_id = match account_id_to_shard_id_map.get(&receipt.receiver_id) {
-                Some(id) => *id,
-                None => {
-                    let id = account_id_to_shard_id(&receipt.receiver_id, shard_layout);
-                    account_id_to_shard_id_map.insert(receipt.receiver_id.clone(), id);
-                    id
-                }
-            };
-            shard_receipts[shard_id as usize].1.push(receipt);
+        let mut cache = HashMap::new();
+        let mut result = HashMap::with_capacity(num_shards);
```

Comment:
assuming that we have way more receipts than shards, I am not sure how much of an improvement we see here by using `with_capacity()` instead of `new()`.  I am tempted to use just `new()` here.